### PR TITLE
add debugging js overview and get started

### PR DESCRIPTION
**Pull request by kaycebasques**
@Meggin please review

### CODE REVIEW

File: src/content/en/tools/chrome-devtools/javascript/get-started.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+[resume]: /web/tools/chrome-devtools/images/resume-script-execution.png
+
+## Step 3: Step through the code
+
+Now that your code is paused, you can step through it, one line at a time, until
+you discover the line of code that is causing the bug.
+
+1. Click  **Step into next function call** ![Step into next function
+   call][into]{:.devtools-inline} to step through the execution of the
+   `onClick` function, one line at a time. DevTools highlights the
```
Comment (jpmedley) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: Make sure your function names end in '()'.

File: src/content/en/tools/chrome-devtools/javascript/get-started.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+   values as strings, rather than adding them as numbers.
+
+[into]: /web/tools/chrome-devtools/images/step-into.png
+[out]: /web/tools/chrome-devtools/images/step-out.png
+[over]: /web/tools/chrome-devtools/images/step-over.png
+
+## Step 4: Apply a fix
+
+You've identified a potential cause for the bug. All that's left is to test out
+your fix by editing the code and then re-running the demo. You don't
```
Comment (jpmedley) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: This sentence makes it sound like you're skipping the fix and going straight to the test. Then there's a fix in step 1. Can you please take another run at this transition.

File: src/content/en/tools/chrome-devtools/javascript/get-started.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+within the DevTools UI.
+
+1. In the code editor on the **Sources** panel of DevTools, replace
+   `var sum = addend1 + addend2` with
+   `var sum = parseInt(addend1) + parseInt(addend2);` to fix the bug. The
+   `parseInt` function converts strings to integers.
+1. Press <kbd>Command</kbd>+<kbd>S</kbd> (Mac) or
+   <kbd>Control</kbd>+<kbd>S</kbd> (Windows, Linux) to save your change.
+   The background of the code changes to red to indicate that the script has
+   been changed within DevTools.
```
Comment (jpmedley) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: Did I miss where you set up DevTools to save changes to the original?

File: src/content/en/tools/chrome-devtools/javascript/overview.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+description: An overview of debugging JavaScript using Chrome DevTools.
+
+{# wf_updated_on: 2016-11-04 #}
+{# wf_published_on: 2016-11-04 #}
+
+# Overview of Debugging JavaScript in Chrome DevTools {: .page-title }
+
+{% include "web/_shared/contributors/kaycebasques.html" %}
+
+This page provides a conceptual overview of what it means to debug JavaScript
```
Comment (jpmedley) [inline src/content/en/tools/chrome-devtools/javascript/overview.md]: 'what it means to debug' I'm worried this phrase might be confusing to non-native readers.

File: src/content/en/tools/chrome-devtools/javascript/overview.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+### Step 1: Reproduce the bug
+
+Find a way to consistently reproduce the bug. If somebody else reported the
+bug and you can't consistently reproduce it, get more information until you can.
+
+Once you can reproduce it, eliminate any steps that are not relevant to
+reproducing the bug. For example, if the bug report says "scroll up and down
+five times and then click this button", but you can reproduce the bug by just
+clicking the button without scrolling, then scrolling is not needed to
+reproduce the bug. Eliminating unnecessary steps like this helps you find
```
Comment (jpmedley) [inline src/content/en/tools/chrome-devtools/javascript/overview.md]: Delete: 'like this'.

File: src/content/en/tools/chrome-devtools/javascript/overview.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+  line of code breakpoints, too.
+
+Ideally, you want to set your breakpoint somewhere close to the code that's
+causing the bug, so that you don't need to step through a bunch of code
+that's not related to your bug. The breakpoint needs to be set on a line of
+code before the bug, so that you
+can inspect what's happening in your code leading up to the bug. However,
+since you don't know what code is causing the bug, setting breakpoints is
+somewhat of a guessing game. It gets easier with practice.
+
```
Comment (jpmedley) [inline src/content/en/tools/chrome-devtools/javascript/overview.md]: When I don't have a good guess, I set random breakpoints at smaller and smaller halves until I narrow it down. Let me know if you want a complete explanation.

File: src/content/en/tools/chrome-devtools/javascript/overview.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+value, or your script is executing in an unexpected order. You can apply
+a fix to your script directly within the DevTools UI, and then re-run your
+page to see if your change fixed the bug. If it did, remember to commit
+your change to your source code repository!
+
+### Step 5: Repeat, if necessary
+
+Debugging is a trial and error process. You might need to set breakpoints
+and walk through your code multiple times before you find a bug. Or, you
+might need to different possible reasons for why you're page is failing,
```
Comment (jpmedley) [inline src/content/en/tools/chrome-devtools/javascript/overview.md]: Change you're to your.

File: src/content/en/tools/chrome-devtools/javascript/overview.md [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+and then test out each hypothesis individually.
+
+If you've tried different approaches and you still can't find the bug,
+sometimes it helps to reason out loud to yourself about what you expect
+your app to do as opposed to what it's actually
+doing. Seasoned programmers call this [rubber duck
+debugging](https://en.wikipedia.org/wiki/Rubber_duck_debugging). Other times,
```
Comment (jpmedley) [inline src/content/en/tools/chrome-devtools/javascript/overview.md]: Seasoned programmers? I'm not sure about that. I never heard the term until Meggin came back from Grace Hopper.

**Comment by Meggin** [comment]
Copying in review comments (also sent these to Kayce in a Google doc):

## General

I cover these general comments more specifically in detailed review of page; just wanted to highlight them here.

- Suggestion to combine Overview and Get Started into one Get Started guide. You can cover the what and the how inside.
- Suggestion to specifically tie in debugging tasks with the DevTools panel as part of introducing debugging. This might be something that’s done in the intro to panels too, but I want to mention it here, as it should be iterated consistently-- what you are doing when you debug, and the specific panels that you use to do this.
- Suggestion to combine ‘Set Breakpoints’ with ‘How to step through code’ as one doc on breakpoint types and breakpoint UI.

## Overview

- I genuinely like the debugging workflow section. I think it would be even awesomer if you covered an example and show an image/animated gif in the Dev Tools, specifically calling out the Dev Tools panels, etc., that are being used as part of that workflow step.
- We talked about this a bit-- so main feedback is that Overview section can go away. There will be a main Get Started doc that is focused on each step in the debugging workflow.
- Most of the concepts in this overview will carry over nicely to Get Started (there isn’t too much information). The one section that might need some tweaking, is the breakpoints one. But luckily enough, there’s a Set Breakpoints doc that can change-- this can be focused on when you’d set the different break point types and how to do it. Should map nicely.
- All the steps in debugging workflow will move up in Get started doc! They will be the primary focus of the content. And it’s kind of cool, because you’ve mapped this already in the two pages :).
- Debugging basics could be the intro to the Get Started doc-- move it in and re-work to get things as lean as possible.
- For sure, under Step 2: Step and trigger a breakpoint, the paragraph about ‘You set’ … leading into bullet list could be leaned, and point to the doc on setting breakpoints. It’ll be an easy enough section to lean, so you can jump right into the get started example.
- For sure, under stepping through the code, we should explain how to check values are as expected (through a common example), and also how to check expected order (simple example).
- Under Apply a fix, we want to make sure we’ll want to link off to more information on how to save to src code (Dev Tools provides support for this-- source maps, I think?)

## Get Started

- The debugging demo is so awesome. Couple of important bits of feedback on this awesome demo:
I’d suggest to move the instructions to open Dev Tools to this first step, as it’s always good to work in Dev Tools from the start.
You are gonna need to make the demo responsive, as many developers will have their default Dev Tools set to Device Mode. Make sure it’s easy to enter values on the smaller screen and see results.
- In Pause the Code-- you could change number one to be ‘Open Dev Tools if you haven’t already…’ a reminder, in case they didn’t in the previous step.
- Also in Pause the Code-- maybe change auxclick to something else, as I’m not seeing that in current Dev Tools?
- For some reason, Step into next function call (again) isn’t enabled. Might be a bug in demo, or something weird in my version of Dev Tools. Trying again to reproduce. I can force it to go to next step by using skip. So that’s what I’ve done, but then the values are off. Something isn’t quite right in execution, so good to check this.
- Summary of main feedback for this:
Make Get Started the single doc, bringing together Overview steps for debugging workflow, and the ‘how-to’ for those steps.
Make sure the demo is responsive and works. I hit a snag-- so maybe we work together to see what’s up? Could be something I’m doing wrong for sure, or could be something small in terms of our versions of things, etc.
- Love ‘Show me all the breakpoints’ button :).

## How to set breakpoints

- So I’m pretty sure the link to ‘how to step through your code’ might want to point instead to the get started doc. And the sentence will need to change. Basically, the developer might want to work through get started before diving into the specifics on breakpoints. This is different than ‘how to step through code’.
- As I work through this doc, I’m wondering if maybe you could bring together the breakpoint type content with the breakpoint UI (in ‘how to step through code) as one comprehensive guide on the tool UI and the breakpoint types. Instead of trying to write this, I think it’s cool to throw the idea out there, and see what you make of it :). Could be a very cool guide actually.
- I’m gonna stop with the review now, as I think this is a fairly hefty chunk of work for this iteraction :).

**Comment by petele** [comment]
@kaycebasques - when this is ready to go and you're happy with it, please remove the DO NOT MERGE tag.

File: src/content/en/tools/chrome-devtools/javascript/get-started.md (hunk 1 above)
Comment (kaycebasques) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: ✔️

File: src/content/en/tools/chrome-devtools/javascript/get-started.md (hunk 2 above)
Comment (kaycebasques) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: ✔️

File: src/content/en/tools/chrome-devtools/javascript/get-started.md (hunk 3 above)
Comment (kaycebasques) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: There's a paragraph at the bottom now that clarifies what's going on

**Comment by kaycebasques** [comment]
Hey @Meggin, I merged the overview material into the get started. Plz review

Not planning on including the following work in this PR:

* Breakpoints doc updates
* Code stepping doc updates

**Comment by Meggin** [comment]
Hey, Kayce

Have the new changes been merged? I don't think I'm seeing them, but then again, I could just be rusty after the break!

**Comment by kaycebasques** [comment]
@Meggin yes the changes are here. I just rebased the new commits onto the original one.

File: src/content/en/tools/chrome-devtools/_toc.yaml [hunk 9]
```diff
@@ -49,12 +51,14 @@ toc:
         path: /web/tools/chrome-devtools/remote-debugging/webviews
     - title: Inspect and Debug JavaScript
       section:
+      - title: Overview
```
Comment (Meggin) [inline src/content/en/tools/chrome-devtools/_toc.yaml]: Are you still planning on removing this section (as part of combining Overview into Get Started)?

File: src/content/en/tools/chrome-devtools/javascript/get-started.md [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+# Get Started with Debugging JavaScript in Chrome DevTools {: .page-title }
+
+{% include "web/_shared/contributors/kaycebasques.html" %}
+
+This step-by-step, interactive tutorial teaches you the basic
+workflow for debugging JavaScript in Chrome DevTools. The tutorial shows you
+how to debug one specific issue, but the general workflow you learn is helpful
+for debugging all types of JavaScript bugs.
+
+If you're using `console.log()` to find and fix bugs in your code, consider
```
Comment (Meggin) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: Not sure about this second paragraph-- the tutorial later explains why you might want to use different alternatives to console log, so maybe it isn't necessary?

File: src/content/en/tools/chrome-devtools/javascript/get-started.md [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+   until DevTools highlights the line of code above. The highlighted line
+   represents the line of code that you're currently paused on.
+
+When you checked the **click** checkbox, you set up an event-based breakpoint on
+all `click` events. When *any* node is clicked, and that node has a `click`
+handler, DevTools automatically pauses on the first line of that node's
+`click` handler.
+
+Note: This is just one of the many types of breakpoints that DevTools offers.
+What breakpoint you should use depends on what type of issue you're debugging.
```
Comment (Meggin) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: Link to the breakpoints doc.

File: src/content/en/tools/chrome-devtools/javascript/get-started.md [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+   call][over]{:.devtools-inline}. DevTools executes `inputsAreEmpty()`
+   without stepping into it. Notice how DevTools skips a few lines of code.
+   This is because `inputsAreEmpty()` evaluated to false, so the `if`
+   statement's block of code didn't execute.
+
+That's the basic idea of stepping through code. If you look at the code in
+`get-started-1.js`, you can see that the bug is probably somewhere in the
+`updateLabel()` function. Rather than stepping through every line of code,
+you can use another type of breakpoint to pause the code closer to the
+probably location of the bug.
```
Comment (Meggin) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: Remove 'probably'. You could change it to 'probable', but I think removing it all together is a nicer alternative.

File: src/content/en/tools/chrome-devtools/javascript/get-started.md [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+
+Line-of-code breakpoints are the most common type of breakpoint. When
+you've got a specific line of code that you want to pause on, use a
+line-of-code breakpoint. Try it now:
+
+1. Look at the last line of code in `updateLabel()`, which looks like this:
+
+       `label.textContent = addend1 + ' + ' + addend2 + ' = ' + sum;`
+
+1. To the left of this code you can see the line number of this particular
```
Comment (Meggin) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: Missing a comma after 'To the left of this code'.

File: src/content/en/tools/chrome-devtools/javascript/get-started.md [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+As suspected, `sum` is being evaluated as a string, when it should be a
+number. This is the cause of the demo's bug.
+
+A second DevTools alternative to `console.log()` is the Console. Use the
+Console to evaluate arbitrary JavaScript statements.
+Developers commonly use the Console to overwrite the variable values
+when debugging. In your case, the Console can help you test out potential
+fixes for the bug you just discovered. Try it now:
+
+1. Press <kbd>Escape</kbd> to open the Console drawer. It opens at the bottom
```
Comment (Meggin) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: Some developers may already have the Console log open. So maybe reword this slightly to recognize this possibility.

Something like:  If the Console drawer isn't open, press Escape to open it at the bottom of your DevTools window.

File: src/content/en/tools/chrome-devtools/_toc.yaml (hunk 9 above)
Comment (kaycebasques) [inline src/content/en/tools/chrome-devtools/_toc.yaml]: Yup. Fixed.

File: src/content/en/tools/chrome-devtools/javascript/get-started.md (hunk 11 above)
Comment (kaycebasques) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: So, my reason for not linking to the breakpoints doc here is, I want people to complete the tutorial. Having a link here invites the reader to open that link and get distracted by the other doc, reducing their chances of completing this tutorial. Thoughts?

File: src/content/en/tools/chrome-devtools/javascript/get-started.md (hunk 10 above)
Comment (kaycebasques) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: Here I'm trying to appeal to a specific developer audience. Devs who use `console.log()` often don't know that there are better alternatives, so I want to call attention to that right away. "Oh, that's me! I use `console.log()`. Now I'm curious to see what alternatives I have." Thoughts?

File: src/content/en/tools/chrome-devtools/javascript/get-started.md (hunk 12 above)
Comment (kaycebasques) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: Fixed

File: src/content/en/tools/chrome-devtools/javascript/get-started.md (hunk 13 above)
Comment (kaycebasques) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: Fixed

File: src/content/en/tools/chrome-devtools/javascript/get-started.md (hunk 14 above)
Comment (kaycebasques) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: Fixed

Comment (kaycebasques) [inline src/content/en/tools/chrome-devtools/javascript/get-started.md]: Fixed

**Comment by kaycebasques** [comment]
Hey @Meggin I updated per your feedback.

If it LGTY can you [approve your review](https://help.github.com/articles/approving-a-pull-request-with-required-reviews/) so that I can merge? I'll merge myself. I want to clean up the git history before merging.
